### DISCUSSION TITLE
UserRegistration seems not properly encapsulated.

### DISCUSSION THREAD

**Discussion by nalkhovikleverx** [aliaksei-kosyrau-leverx/GoEdu#5] [Ideas]
# Problem
The invariants are meaningless in conditions of poor encapsulation. Right now you can edit the state of UserRegistration model from the lowest layers. It puts the model into an invalid state where model invariants aren't controlled.

# Analysis
The question is how to maintain the encapsulation of UserRegistration in Go when we need its internal state to form a database query, or vice versa, when we need to restore the state of the UserRegistration from the database.

# Possible Solution
To keep things simple we can use some sort of DTO which crosses not a network boundaries, but a logical boundaries of a model. We can use this DTO to populate the database state. Also we can use it to restore the model from the database as well.

The similar example we can see in:
* [EventStore reference application](https://github.com/EventStore/training-advanced-go/blob/main/domain/doctorday/day.go#L225)
* [Modular Monolith with DDD by Kamil Grzybek](https://github.com/kgrzybek/modular-monolith-with-ddd/blob/6f707973ac33672890f369b444adf2a982cf510a/src/Modules/Payments/Domain/SubscriptionPayments/SubscriptionPaymentSnapshot.cs#L6)

Other solution is demonstrated in following projects:
* [Wild Workouts By ThreeDots](https://github.com/ThreeDotsLabs/wild-workouts-go-ddd-example/blob/master/internal/trainings/domain/training/training.go#L51)

**Comment by aliaksei-kosyrau-leverx** [comment]
Ok, as I understand it we need to add constructors for our basic models and rename the fields with a small letter to make them private, is this correct? Add getters/setters for private fields

**Reply by nalkhovikleverx** [reply]
To keep our `UserRegistration` struct encapsulated we must definitely make all its fields private and left only public methods. 

Instead of using "getters" we can define a new struct type called `UserRegistrationSnapshot` which will hold a copy of all UserRegistration model private fields. These fields must be public without any methods. In my opinion this approach will work best for us because we can easily save the snapshot to the database and later restore it. 

To build the `UserRegistration` from `UserRegistrationSnapshot` we should define a new constructor for `UserRegistration` that will map it from `UserRegistrationSnapshot`. This constructor should use `Must...` functions to build all VO because they must be valid.
